add dev command

BlaineHeffron opened the pull request:
adds dev command

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,103 @@
+            .manifest_path
+            .parent()
+            .map(|path| path.to_path_buf())
+            .unwrap_or_else(|| PathBuf::from("."));
```
You should be able to do `Path::new(".")` and avoid needing the map. `workspace_root` can be a `Path`

BlaineHeffron replied:
got it

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,103 @@
+            .map(|path| path.to_path_buf())
+            .unwrap_or_else(|| PathBuf::from("."));
+        let env_toml = workspace_root.join("environments.toml");
+        if env_toml.exists() {
```
Does watching a non-existent file cause an error? Otherwise it would seem like you would want to watch for the files creation as an update.

BlaineHeffron replied:
yes, `error: No such file or directory`

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,103 @@
+                if let Ok(event) = res {
+                    if let notify::EventKind::Modify(_) = event.kind {
+                        if let Some(path) = event.paths.first() {
+                            println!("File modified: {:?}", path);
```
```suggestion
                            eprintln!("File modified: {:?}", path);
```

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,103 @@
+        let tx_clone = tx.clone();
+        let mut watcher =
+            notify::recommended_watcher(move |res: Result<notify::Event, notify::Error>| {
+                if let Ok(event) = res {
```
```suggestion
                if let Ok(event @ notify::Event { kind: notify::EventKind::Modify(_) } ) = res {
```
Since the assignment is a pattern match you can match on the inner type and the `@` lets you simultaneously assign the value to a variable. So this has the same effect as the nested ifs. See example here:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a2ae21a539c81d4d157644d6da6b978b

willemneal replied:
Also this suggestion is not good on it's own since it doesn't remove the inner if.

willemneal replied:
Also since these types don't collide with any in this file I'd just import them for better readability.

BlaineHeffron replied:
done

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,103 @@
+        // Set LOAM_ENV to development
+        std::env::set_var("LOAM_ENV", "development");
+        let (tx, mut rx) = tokio::sync::mpsc::channel(100);
+        let tx_clone = tx.clone();
```
What errors were you getting if this wasn't cloned?

BlaineHeffron replied:
It works if it isn't cloned. There would be  a problem if tx was used later in the code but since it isn't I don't think we need to clone it.

willemneal commented on crates/loam-cli/src/commands/mod.rs:
```diff
@@ -39,6 +40,7 @@ impl Root {
         match &mut self.cmd {
```
Ah so since this a mutable reference you can change `dev_info.run` to use `(&mut self)`, which removes the need to clone in the `build` method.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,103 @@
+            .build_clients
+            .env
+            .get_or_insert(LoamEnv::Development);
+        build_cmd.profile = "dev".to_string();
```
@chadoh Do you ever expect someone to use a non-dev profile? Currently the default is set by clap, but we could change to an Option like `env` which would let us respect what the user pasted in and not overwrite it here.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,103 @@
+            .build_clients
+            .env
+            .get_or_insert(LoamEnv::Development);
+        build_cmd.profile = "dev".to_string();
```
```suggestion
        build_cmd.profile = "debug".to_string();
```

BlaineHeffron replied:
Seems to not like using the name debug: 
```
error: profile name `debug` is reserved
To configure the default development profile, use the name `dev` as in [profile.dev]
```

willemneal replied:
Ah yeah. `cargo build` is debug by default, you need to add either `--release` or a `--profile <defined in Cargo.toml>`. So this is more reason to make this an option and then it will be debug by default.

BlaineHeffron replied:
I see. I changed it to be an option and default to debug. If its debug it doesnt add the --profile command.

chadoh commented on crates/loam-cli/src/commands/build/mod.rs:
```diff
@@ -125,10 +131,12 @@ impl Cmd {
-            if self.profile == "release" {
+            let profile = self.profile.clone().unwrap_or("release".to_string());
+            if profile == "release" {
                 cmd.arg("--release");
+            } else if profile == "debug" {
             } else {
-                cmd.arg(format!("--profile={}", self.profile));
+                cmd.arg(format!("--profile={profile}"));
             }
```
if it is neither `release` nor `debug` then we ignore it?

willemneal replied:
No we ignore if it is debug. Or said another way don't ignore it if isn't debug.

chadoh commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,105 @@
+impl Cmd {
+    pub async fn run(&self) -> Result<(), Error> {
+        // Set LOAM_ENV to development
+        std::env::set_var("LOAM_ENV", "development");
```
I think we should _default_ `LOAM_ENV` to `development` _when you use `loam dev`_. And when you use `loam build`, we should default it to `production`. But you should be able to override the defaults.

I think this is the behavior of other commands like `astro dev` and `astro build`.

If you want to run the watcher but compile everything for `LOAM_ENV=production`, why should we stop you?

BlaineHeffron replied:
Good catch, I should have deleted this line. It is currently defaulting to development later when it runs build.

chadoh commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,105 @@
+            .parent()
+            .unwrap_or_else(|| Path::new("."));
+        let env_toml = workspace_root.join("environments.toml");
+        if env_toml.exists() {
```
In addition to noticing if you _modify_ `environments.toml`, can we also check if you _add_ or _delete_ it?

BlaineHeffron replied:
I modified the code to do this. It was a pretty substantial change because often `environments.toml` is in a  directory not otherwise watched. To do this I had to watch its parent directory but then make sure only changes to that file were made (assuming its not in a directory already being watched for changes). I used Arc<PathBuf> and Arc::clone to do this because the lifetimes of the paths being passed to the watcher function were causing me issues.

chadoh reviewed (changes requested):
Some questions/comments. Also, I want to see some tests!

willemneal commented on crates/loam-cli/src/commands/build/mod.rs:
```diff
@@ -125,10 +131,12 @@ impl Cmd {
+            let profile = self.profile.clone().unwrap_or("release".to_string());
+            if profile == "release" {
                 cmd.arg("--release");
+            } else if profile == "debug" {
             } else {
```
```suggestion
            } else if profile != "debug" {
```

willemneal commented on crates/loam-cli/src/commands/build/mod.rs:
```diff
@@ -125,10 +131,12 @@ impl Cmd {
             cmd.arg("--crate-type=cdylib");
             cmd.arg("--target=wasm32-unknown-unknown");
-            if self.profile == "release" {
+            let profile = self.profile.clone().unwrap_or("release".to_string());
```
You can use `.as_ref` on an option to go from `Option<T>` to `Option<&T>` and in this case you can use `as_deref` would would go `Option<String>` to `Option<&str>` preventing the need to clone and use a string.
Second you should have used `unwrap_or_else` because this way you call `release.to_string` even if the `Option` is a `Some`.

```suggestion
            let profile = self.profile.as_deref().unwrap_or("release");
```

BlaineHeffron commented:
Setting to draft while I working on watching potential additions / deletions of environments.toml and adding unit tests.

BlaineHeffron commented:
Some substantial changes have been made. First, I found out while testing that the debounce mechanism wasn't working properly. A mutex state is used to track whether or not rebuilding should be blocked. This makes it so that if multiple files are changed while the build process is ongoing, it will only trigger one rebuild at the end. Before it would trigger N rebuilds where N were the number of changes detected.

Second, the logic for determining changes was overhauled. File modifications, deletions, and insertions in the watched directories all trigger rebuilds. Same with `environments.toml`. I needed to ensure that files in the same directory as `environments.toml` are ignored (this is because now it must watch the parent directory of `environments.toml` to determine if creations/deletions are made), so I needed to use `Arc` to pass the path objects into the watcher.

Finally, I added tests. They are all in one test block because the initial build of the directory takes some time. This modifies environments.toml and makes sure its loading in the new configuration. It also changes a file within one of the contracts directory and ensures it is detected and rebuilt.

chadoh commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,208 @@
+    pending: bool,
+}
+
+enum Message {
+    FileChanged,
+}
```
an enum with one variant seems odd

BlaineHeffron replied:
fixed

chadoh commented on crates/loam-cli/tests/it/build_clients/dev.rs:
```diff
@@ -0,0 +1,142 @@
+                "🌐 using network at http://localhost:9000/rpc",
+            )
+            .await;
+            dev_process
+                .kill()
+                .await
+                .expect("Failed to kill dev process");
+
+            dev_process
+                .wait()
+                .await
+                .expect("Failed to wait for dev process");
```
What's happening here?

BlaineHeffron replied:
oops, good eye. I removed the unnecessary call at the end.

chadoh reviewed (approved):
Wow so glad I asked for tests! Two small questions below. I also wonder if it's worth testing the "environments.toml added/removed" cases.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,208 @@
+use super::build::clients::LoamEnv;
+
+struct RebuildState {
+    pending: bool,
```
Similar question as chad here. Is this needed or can it just be a bool?

BlaineHeffron replied:
changed to just a bool

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,208 @@
+        }
+        println!("Watching for changes. Press Ctrl+C to stop.");
+
+        let rebuild_state_clone = Arc::clone(&rebuild_state);
```
```suggestion
        let rebuild_state_clone = rebulid_state.clone();
```
Does this not work?

willemneal replied:
Also do you need this at all? Can't you just `rebuild_state.clone()` below each time?

BlaineHeffron replied:
Yes, I changed it, thanks.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,208 @@
+}
+
+impl Cmd {
+    pub async fn run(&self) -> Result<(), Error> {
```
```suggestion
    pub async fn run(&mut self) -> Result<(), Error> {
```

willemneal replied:
We can do this because the caller has a mutable reference! Then you can do.
```
         self.build_cmd
            .build_clients
            .env
            .get_or_insert(LoamEnv::Development);
        self.build_cmd.profile.get_or_insert_with(|| "debug".to_string());
        let build_cmd = Arc::new(build_cmd.clone());
```

Then below you can replace `self.clone()` with `build_cmd.clone()`, which just increments the counter and creates a new reference.

BlaineHeffron replied:
Okay, refactored to only clone build_cmd. Let me know what you think.

BlaineHeffron replied:
I reverted the changes because it was causing a failure during the build command. I'm not sure why what I did didn't work. If you think its important I can make another attempt at it.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,208 @@
+        // Debounce to avoid multiple rapid rebuilds
+        time::sleep(std::time::Duration::from_secs(1)).await;
+
+        println!("Changes detected. Rebuilding...");
```
```suggestion
        eprintln!("Changes detected. Rebuilding...");
```

BlaineHeffron replied:
should all the println statements be eprintln?

willemneal commented on crates/loam-cli/tests/it/build_clients/dev.rs:
```diff
@@ -0,0 +1,142 @@
+        }
+    })
+    .await;
+    match result {
+        Ok(_) => {
+            println!("Found string {}", expected)
+        }
+        Err(_) => panic!("Timed out waiting for output: {}", expected),
+    }
```
```suggestion
    match result {
        Ok(_) => {
            println!("Found string {expected}")
        }
        Err(_) => panic!("Timed out waiting for output: {expected}"),
    }
```

BlaineHeffron replied:
It doesn't work, doesnt this require a string format call or something?

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -143,7 +135,7 @@ impl Cmd {
                             {
                                 println!("File changed: {path:?}");
-                                if let Err(e) = tx.blocking_send(Message::FileChanged) {
+                                if let Err(e) = tx.blocking_send("FileChanged".to_string()) {
```
I actually prefer the enum with a single case over a string. For one it's less to send.

BlaineHeffron replied:
Okay, I'll go with it.

BlaineHeffron commented:
It has been refactored to only clone the build command. I also fixed an issue where environments.toml deletions were not being detected. I added a test case for that. All comments have been addressed.

chadoh reviewed (approved):
Heroic

chadoh commented:
Go ahead and get the PR title and description, or maybe the squash commit message (you'll have to check the set up) into a format that can be used by the auto release logic. And then merge away!

willemneal commented on crates/loam-cli/src/commands/build/clients.rs:
```diff
@@ -62,8 +69,8 @@ impl Args {
-    fn loam_env(self) -> String {
-        format!("{:?}", self.env).to_lowercase()
+    fn loam_env(self, default: LoamEnv) -> String {
+        self.env.unwrap_or(default).to_string()
```
```suggestion
        self.env.unwrap_or(default).to_lowercase()
```
It returns a `String` and we still want to make sure it's lowercase.

BlaineHeffron replied:
good catch

willemneal commented on crates/loam-cli/src/commands/build/mod.rs:
```diff
@@ -92,11 +92,17 @@ pub enum Error {
+    pub fn list_packages(&self) -> Result<Vec<Package>, Error> {
         let metadata = self.metadata()?;
         let packages = self.packages(&metadata)?;
         let packages = loam_build::deps::get_workspace(&packages)?;
+        Ok(packages)
```
```suggestion
        Ok(loam_build::deps::get_workspace(&packages)?)
```

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,217 @@
+                            if is_temporary_file(path) {
+                                return;
+                            }
+                            let env_toml_parent_abs = canonicalize_path(&env_toml_parent_clone);
+                            let env_toml_path_abs = canonicalize_path(&env_toml_path_clone);
```
Can this be done outside?

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,217 @@
+    let file_name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
+
+    // Vim temporary files
+    if file_name.starts_with('.') {
```
This could also be an `.env` file. Do they end with `.swp`?

BlaineHeffron replied:
`.swp` wasn't the only one but I think `.swo`, `~` and `.swp` should cover it all.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,217 @@
+        return true;
+    }
+    // VSCode temporary files
+    if std::path::Path::new(file_name)
```
```suggestion
    if Path::new(file_name)
```

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,217 @@
+                            let parent_is_env_toml_parent =
+                                path.parent() == Some(env_toml_parent_abs.as_path());
+                            let path_is_env_toml = path == env_toml_path_abs.as_path();
+                            let parent_is_in_watched_dirs = is_parent_in_watched_dirs(
+                                &env_toml_parent_abs,
+                                &watched_dirs_clone,
```
This seems like it could also be computed outside the closure.  I'm I don't understand why this is an issue.

BlaineHeffron replied:
Fixed

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,217 @@
+                                &watched_dirs_clone,
+                            );
+                            // skip if the file is in the parent directory of environments.toml and it is not environments.toml
+                            if !(parent_is_env_toml_parent
+                                && !path_is_env_toml
+                                && !parent_is_in_watched_dirs)
```
```suggestion
                            if !parent_is_env_toml_parent
                                || path_is_env_toml
                                || parent_is_in_watched_dirs
```
It's best to distribute the negation as it's less work for the reader.
So in this case if `parent_is_in_watched_dirs` is true which is independent of the `path` argument this is invoked?

BlaineHeffron replied:
yes because any file in the watched dirs should trigger a rebuild.

BlaineHeffron replied:
basically we have to add the parent dir of the env file to the list of dirs in order to check if there is a deletion or creation of the env file. If that dir is already in the watched dirs then all watched files should trigger a rebuild

BlaineHeffron replied:
I realized the parent_is_env_toml_parent check is redundant so its been removed in the refactored version

willemneal commented:
I have one thought that would help clean this up a bit.

```rust
#[derive(Clone)
struct Watcher {
    root_env: Arc<PathBuf>,
    packages: Arc<Vec<PathBuf>,
}

impl Watcher {
    pub fn new(root_env: &Path, packages: &[PathBuf]) -> Self {
        Self {
            root_env: Arc::new(conanicalize_path(root_env))
            packages: Arc::new(packages.iter().map(canonicalize_path).collect())
        }
    }
/// Relevant methods here, e.g.
    pub fn is_env_toml_parent(path: &Path) -> bool {
        self.root_env.parent() == path.parent()
    }
}

```

Since `Arc` implements `Clone` we can derive it for this struct. Then it can be `.clone()` and the relevant logic placed inside.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+    pub fn new(root_env: &Path, packages: &[PathBuf]) -> Self {
+        Self {
+            root_env: Arc::new(canonicalize_path(root_env)),
+            packages: Arc::new(packages.iter().map(|p| canonicalize_path(p)).collect()),
```
```suggestion
            packages: Arc::new(packages.iter().map(canonicalize_path).collect()),
```

BlaineHeffron replied:
Seems to complain about the type: ``` note: expected function signature `fn(&PathBuf) -> _`
               found function signature `fn(&std::path::Path) -> _`
note: required by a bound in `std::iter::Iterator::map`
   --> /home/blaine/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:764:12
    |
761 |     fn map<B, F>(self, f: F) -> Map<Self, F>
    |        --- required by a bound in this associated function
...
764 |         F: FnMut(Self::Item) -> B,
    |            ^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Iterator::map`
help: consider wrapping the function in a closure```

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+
+fn is_temporary_file(path: &Path) -> bool {
+    const IGNORED_EXTENSIONS: &[&str] = &["tmp", "swp", "swo"];
+    let file_name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
```
The default af `""` seems strange. The reason you need to convert to a string is that a `str` is UTF-8 whereas  a Path uses an OsStr because some OS's can have path names which aren't safely converted to UTF-8. So it's trying to convert to a string and returning Option of whether it's valid UTF-8. We could consider using [camino](https://docs.rs/camino/latest/camino) which is a crate that enforces paths be UTF-8 (which for most rust projects is definitely the case). I've been wanting to try it out because then you can go between a Path and `str` without any headaches.

But for now I'd use an `expect` and we could use the other crate in a follow up PR.  I think it would clean up the whole repo.

BlaineHeffron replied:
Okay, set it to use `expect` for now.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+
+fn is_temporary_file(path: &Path) -> bool {
+    const IGNORED_EXTENSIONS: &[&str] = &["tmp", "swp", "swo"];
+    let file_name = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
```
```suggestion
    let file_name = path.file_name().and_then(Path::to_str).unwrap_or("");
```

BlaineHeffron replied:
`file_name` returns an `Option<&OsStr>` which isn't compatible with `Path::to_str`

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+            eprintln!("Watching {}", package_path.display());
+        }
+
+        let watcher_clone = watcher.clone();
```
Why still the initial clone?

BlaineHeffron replied:
```error[E0382]: borrow of moved value: `watcher`
    --> crates/loam-cli/src/commands/dev/mod.rs:163:29
     |
127  |         let watcher = Watcher::new(&env_toml_path, &packages);
     |             ------- move occurs because `watcher` has type `dev::Watcher`, which does not implement the `Copy` trait
...
135  |             notify::recommended_watcher(move |res: Result<notify::Event, notify::Error>| {
     |                                         ------------------------------------------------ value moved into closure here
...
147  |                             if watcher.is_watched(path) || watcher.is_env_toml(path) {
     |                                ------- variable moved due to use in closure
...
163  |         for package_path in watcher.packages.iter() {
     |                             ^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after move
     |
     = note: borrow occurs due to deref coercion to `Vec<PathBuf>`
note: deref defined here
    --> /home/blaine/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/sync.rs:2102:5
     |
2102 |     type Target = T;
     |     ^^^^^^^^^^^
```

willemneal replied:
Yes but why not clone it inside the closure as I have below?

BlaineHeffron replied:
Looks like I can avoid cloning if I use the package and env paths directly when setting the notify_watcher to watch them.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+                            if is_temporary_file(path) {
+                                return;
+                            }
+                            if watcher_clone.is_watched(path) || watcher_clone.is_env_toml(path) {
```
```suggestion
                            let watcher = watcher.clone();
                            if watcher.is_watched(path) || watcher.is_env_toml(path) {
```

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+            .build_cmd
+            .list_packages()?
+            .into_iter()
+            .map(|package| PathBuf::from(package.manifest_path.parent().unwrap().as_str()))
```
Isn't `parent` an `Option<&Path>`?
```suggestion
            .map(|package| package.manifest_path.parent().unwrap().into_path_buf())
```

BlaineHeffron replied:
Seems the `to_path_buf` method returns a `Utf8PathBuf` which we'd then have to refactor alot of the code to pass that type. Is that what you'd prefer?

willemneal replied:
This is fine for now. Again, we can return with `camino` later to clean up all path stuff.

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+
+        let watcher = Watcher::new(&env_toml_path, &packages);
+
+        for package_path in watcher.packages.iter() {
```
```suggestion
        for package_path in &watcher.packages {
```

BlaineHeffron replied:
This results in the following error: 
```
&Arc<Vec<PathBuf>>` is not an iterator
the trait `Iterator` is not implemented for `&Arc<Vec<PathBuf>>`, which is required by `&Arc<Vec<PathBuf>>: IntoIterator`
required for `&Arc<Vec<PathBuf>>` to implement `IntoIterator
```

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+            watcher.root_env.parent().unwrap(),
+            RecursiveMode::NonRecursive,
+        )?;
+        for package_path in watcher.packages.iter() {
```
```suggestion
        for package_path in &watcher.packages {
```

BlaineHeffron replied:
Same issue as above

willemneal commented on crates/loam-cli/src/commands/dev/mod.rs:
```diff
@@ -0,0 +1,219 @@
+            RecursiveMode::NonRecursive,
+        )?;
+        for package_path in watcher.packages.iter() {
+            notify_watcher.watch(package_path, RecursiveMode::Recursive)?;
```
It would be nice to be able to filter out the temporary files here. But that is an optimization we can deal with later (though I doubt we'll ever need)

BlaineHeffron replied:
I think the difficulty is that its watching the whole directory so that it catches new files, so we couldn't calculate if a new file matches the temporary file pattern a-priori. Maybe there is a method I'm not aware of that does that you propose.

willemneal replied:
Yeah no worries. Again it's filtering later works just as well.

willemneal reviewed (approved):
Great work! A deep dive into Rust!
